File: TeamCodeA/src/main/java/org/firstinspires/ftc/teamcodea/pedroPathing/Paths.java
package org.firstinspires.ftc.teamcodea.pedroPathing;

import com.pedropathing.follower.Follower;
import com.pedropathing.geometry.BezierCurve;
import com.pedropathing.geometry.BezierLine;
import com.pedropathing.geometry.Pose;
import com.pedropathing.paths.PathChain;

public class Paths {

    private final Follower follower;
    private final int routeSerial;

    public PathChain Path1, Path2, Path3, Path4, Path5, Path6, Path7, Path8;

    public Paths(Follower follower, int routeSerial) {
        this.follower = follower;
        this.routeSerial = routeSerial;

        buildPaths();
    }

    private void buildPaths() {

        if (routeSerial == 1) {

            Path1 = follower.pathBuilder()
                    .addPath(new BezierLine(
                            new Pose(21.762, 120.695),
                            new Pose(43.061, 99.704)
                    ))
                    .setConstantHeadingInterpolation(Math.toRadians(135))
                    .build();

            Path2 = follower.pathBuilder()
                    .addPath(new BezierCurve(
                            new Pose(43.061, 99.704),
                            new Pose(78.714, 82.264),
                            new Pose(17.216, 86.239)
                    ))
                    .setTangentHeadingInterpolation()
                    .build();

            Path3 = follower.pathBuilder()
                    .addPath(new BezierLine(
                            new Pose(17.216, 86.239),
                            new Pose(43.215, 99.859)
                    ))
                    .setTangentHeadingInterpolation()
                    .build();

            Path4 = follower.pathBuilder()
                    .addPath(new BezierCurve(
                            new Pose(43.215, 99.859),
                            new Pose(88.592, 55.563),
                            new Pose(16.733, 62.749)
                    ))
                    .setTangentHeadingInterpolation()
                    .build();

            Path5 = follower.pathBuilder()
                    .addPath(new BezierLine(
                            new Pose(16.733, 62.749),
                            new Pose(43.061, 99.550)
                    ))
                    .setTangentHeadingInterpolation()
                    .build();

            Path6 = follower.pathBuilder()
                    .addPath(new BezierCurve(
                            new Pose(43.061, 99.550),
                            new Pose(95.383, 31.177),
                            new Pose(15.285, 39.741)
                    ))
                    .setTangentHeadingInterpolation()
                    .build();

            Path7 = follower.pathBuilder()
                    .addPath(new BezierLine(
                            new Pose(15.285, 39.741),
                            new Pose(43.215, 99.704)
                    ))
                    .setTangentHeadingInterpolation()
                    .build();

            Path8 = follower.pathBuilder()
                    .addPath(new BezierLine(
                            new Pose(43.215, 99.704),
                            new Pose(54.482, 48.000)
                    ))
                    .setTangentHeadingInterpolation()
                    .build();
        }else if (routeSerial ==2){
            Path1 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierLine(new Pose(21.762, 120.695), new Pose(43.061, 99.704))
                    )
                    .setConstantHeadingInterpolation(Math.toRadians(135))
                    .build();

            Path2 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierLine(new Pose(43.061, 99.704), new Pose(48.154, 129.646))
                    )
                    .setTangentHeadingInterpolation()
                    .build();

            Path3 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierLine(new Pose(48.154, 129.646), new Pose(47.383, 35.653))
                    )
                    .setTangentHeadingInterpolation()
                    .build();
        }else if(routeSerial ==3){
            Path1 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierLine(new Pose(122.701, 121.929), new Pose(101.865, 101.402))
                    )
                    .setConstantHeadingInterpolation(Math.toRadians(45))
                    .build();

            Path2 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierCurve(
                                    new Pose(101.865, 101.402),
                                    new Pose(63.125, 79.486),
                                    new Pose(125.325, 82.727)
                            )
                    )
                    .setTangentHeadingInterpolation()
                    .build();

            Path3 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierLine(new Pose(125.325, 82.727), new Pose(101.711, 101.556))
                    )
                    .setTangentHeadingInterpolation()
                    .build();

            Path4 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierCurve(
                                    new Pose(101.711, 101.556),
                                    new Pose(58.804, 51.859),
                                    new Pose(126.559, 58.804)
                            )
                    )
                    .setTangentHeadingInterpolation()
                    .build();

            Path5 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierLine(new Pose(126.559, 58.804), new Pose(101.865, 101.402))
                    )
                    .setTangentHeadingInterpolation()
                    .build();

            Path6 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierCurve(
                                    new Pose(101.865, 101.402),
                                    new Pose(41.672, 30.714),
                                    new Pose(124.707, 34.881)
                            )
                    )
                    .setTangentHeadingInterpolation()
                    .build();

            Path7 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierLine(new Pose(124.707, 34.881), new Pose(101.556, 101.248))
                    )
                    .setTangentHeadingInterpolation()
                    .build();

            Path8 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierLine(new Pose(101.556, 101.248), new Pose(86.122, 54.482))
                    )
                    .setTangentHeadingInterpolation()
                    .build();
        }else if(routeSerial==4){
            Path1 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierLine(new Pose(122.238, 121.929), new Pose(101.865, 101.402))
                    )
                    .setConstantHeadingInterpolation(Math.toRadians(135))
                    .build();

            Path2 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierLine(new Pose(101.865, 101.402), new Pose(91.987, 132.270))
                    )
                    .setTangentHeadingInterpolation()
                    .build();

            Path3 = follower
                    .pathBuilder()
                    .addPath(
                            new BezierLine(new Pose(91.987, 132.270), new Pose(91.061, 34.109))
                    )
                    .setTangentHeadingInterpolation()
                    .build();
        }

        }

        }

File: TeamCodeD/src/main/java/org/firstinspires/ftc/teamcode/opmode/Paths.kt
package org.firstinspires.ftc.teamcode.opmode

import com.pedropathing.follower.Follower
import com.pedropathing.geometry.BezierCurve
import com.pedropathing.geometry.BezierLine
import com.pedropathing.geometry.Pose
import com.pedropathing.paths.PathChain
import org.firstinspires.ftc.teamcode.constants.AutoConstants
import org.firstinspires.ftc.teamcode.utils.Alliance
import kotlin.math.atan2
import kotlin.math.sqrt
import kotlin.properties.Delegates

class Paths(private val follower: Follower) {
    init {
        buildPaths(Alliance.RED)
    }
    lateinit var fromStartToShoot: PathChain
    lateinit var pickUpFirstRow: PathChain
    lateinit var pickUpFirstRowWithGate: PathChain
    lateinit var firstRowToShoot: PathChain
    lateinit var pickUpSecondRow: PathChain
    lateinit var secondRowToShoot: PathChain
    lateinit var pickUpThirdRow: PathChain
    lateinit var startPosition: Pose
    lateinit var shootPosition: Pose
    lateinit var firstRowControlPoint: Pose
    lateinit var firstRow: Pose
    lateinit var secondRowControlPoint: Pose
    lateinit var secondRow: Pose
    lateinit var thirdRow: Pose
    lateinit var thirdRowControlPoint: Pose
    lateinit var thirdRowEndPoint: Pose
    lateinit var endLocation: Pose
    lateinit var innerEndLocation: Pose
    lateinit var goalLocation: Pose
    var aimHeading = 0.0;


    fun buildPaths(alliance: Alliance) {
        fun Pose.mirrorIfBlue(): Pose {
            return if (alliance == Alliance.BLUE) this.mirror() else this
        }

        // Assume robot starts aligned with goal, facing obelisk wall, with line in the middle of the drive train
        // https://www.desmos.com/calculator/qgkrj2eng6
        startPosition = Pose(130.046 - ROBOT_WIDTH / sqrt(2.0) / 2, 130.046 - ROBOT_WIDTH / sqrt(2.0) / 2, Math.toRadians(126.0)).mirrorIfBlue()
        shootPosition = Pose(84.8704156479217, 78.42542787286067).mirrorIfBlue()
        firstRowControlPoint = Pose(91.99073396778475, 83.2123552123552).mirrorIfBlue()
        firstRow = Pose(144 - 6 - ROBOT_LENGTH / 2 - ROBOT_FRONT_PROTRUSION, 83.51).mirrorIfBlue()
        secondRowControlPoint = Pose(83.14671814671814, 54.622779922779934).mirrorIfBlue()
        secondRow = Pose(144 - ROBOT_LENGTH / 2 - ROBOT_FRONT_PROTRUSION - 6, 54.7 + 2.0).mirrorIfBlue()
        thirdRow = Pose(98.30656370656376, 35.80328185328186).mirrorIfBlue()
        thirdRowControlPoint = Pose(83.03667953667953, 35.43976833976834).mirrorIfBlue()
        thirdRowEndPoint = Pose(144 - ROBOT_LENGTH / 2 - ROBOT_FRONT_PROTRUSION - 6, 35.43976833976834).mirrorIfBlue()

        endLocation = Pose(100.0, 53.0).mirrorIfBlue()
        innerEndLocation = Pose(86.0, 132.0).mirrorIfBlue()
        val pickupHeading = if (alliance == Alliance.RED) 0.0 else Math.toRadians(180.0)


        goalLocation = Pose(138.0, 144.0).mirrorIfBlue()
        aimHeading = calculateAimHeading(shootPosition, goalLocation)

        fromStartToShoot = follower.pathBuilder().addPath(
            BezierLine(startPosition, shootPosition)
        ).build()

        pickUpFirstRow = follower.pathBuilder()
            .addPath(BezierCurve(shootPosition, firstRowControlPoint, firstRow))
            .setConstantHeadingInterpolation(pickupHeading)
            .setReversed()
            .addPoseCallback(Pose(101.749, 82.334).mirrorIfBlue(), {
                follower.setMaxPower(AutoConstants.MAX_POWER_WHEN_RUNNING_INTAKE)
            }, 0.5)
            .build()

        pickUpFirstRowWithGate = follower.pathBuilder()
            .addPath(BezierCurve(shootPosition, firstRowControlPoint, firstRow))
            .setConstantHeadingInterpolation(pickupHeading)
            .setReversed()
            .addPoseCallback(Pose(101.749, 82.334 - 12.0).mirrorIfBlue(), {
                follower.setMaxPower(AutoConstants.MAX_POWER_WHEN_RUNNING_INTAKE)
            }, 0.5)
            .build()

        firstRowToShoot = follower.pathBuilder()
            .addPath(BezierLine(firstRow, shootPosition))
            .setLinearHeadingInterpolation(pickupHeading, aimHeading)
            .build()

        pickUpSecondRow = follower.pathBuilder()
            .addPath(BezierCurve(shootPosition, secondRowControlPoint, secondRow))
            .setConstantHeadingInterpolation(pickupHeading)
            .setReversed()
            .addPoseCallback(Pose(97.773, 60.0).mirrorIfBlue(), {
                follower.setMaxPower(AutoConstants.MAX_POWER_WHEN_RUNNING_INTAKE)
            }, 0.5)
            .build()

        secondRowToShoot = follower.pathBuilder()
            .addPath(BezierCurve(secondRow, secondRowControlPoint, shootPosition))
            .setLinearHeadingInterpolation(pickupHeading, aimHeading)
            .build()

        pickUpThirdRow = follower.pathBuilder()
            .addPaths(
                BezierCurve(shootPosition, thirdRowControlPoint, thirdRow),
                BezierLine(thirdRow, thirdRowEndPoint)
            )
            .setGlobalTangentHeadingInterpolation()
            .setGlobalReversed()
            .build()
    }

    companion object {
        fun calculateAimHeading(robot: Pose, goal: Pose): Double {
            return atan2(goal.y - robot.y, goal.x - robot.x)
            // alternate implementation
            // goal.minus(robot).asVector.theta
        }
        val obelisk = Pose(72.0, 144.0)
        const val ROBOT_LENGTH = 17.055;
        const val ROBOT_WIDTH = 16.850394;
        const val ROBOT_FRONT_PROTRUSION = 9.0;
        val RED_GOAL_FRONT = Pose(130.0, 130.0)
        val BLUE_GOAL_FRONT = Pose(14.0, 130.0)
    }
}